### PR TITLE
rpm: add python3-saml as install dependency

### PR DESCRIPTION
**Pull request by epuertat**
`python.*-saml` is required for Ceph-Dashboard SSO support (optional
feature).

Fixes: https://tracker.ceph.com/issues/44721
Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: debian/control [hunk 1]
```diff
@@ -251,6 +251,7 @@ Depends: ceph-mgr (= ${binary:Version}),
          ${misc:Depends},
          ${python:Depends},
          ${shlibs:Depends},
+Suggests: python3-saml
```
Comment (tchaikov) [inline debian/control:254]: i cannot find python3-saml in either debian or ubuntu official repo:

- https://packages.debian.org/search?keywords=python3-saml&searchon=names&suite=all&section=all
- https://packages.ubuntu.com/search?keywords=python3-saml&searchon=names&suite=all&section=all

File: ceph.spec.in [hunk 2]
```diff
@@ -544,6 +544,7 @@ Requires:       python%{python3_pkgversion}-PyJWT
 Requires:       python%{python3_pkgversion}-Routes
 Requires:       python%{python3_pkgversion}-Werkzeug
 %endif
+Suggests:     python%{python3_pkgversion}-saml
```
Comment (tchaikov) [inline ceph.spec.in:547]: it's available in fedora and openSUSE Tubleweed though. but with different names. see https://pkgs.org/search/?q=python3-saml

i am not sure what "python2dist(python3-saml)" implies. if it means "python3-saml", it's okay. otherwise could update accordingly?

File: debian/control (hunk 1 above)
Comment (epuertat) [reply debian/control:254]: Yup, I know it, I checked it yesterday :/ As it's a weak dependency, it shouldn't block the install, right?

File: ceph.spec.in (hunk 2 above)
Comment (epuertat) [reply ceph.spec.in:547]: Thanks for pointing that out, @tchaikov! Will fix it.

File: ceph.spec.in [hunk 3]
```diff
@@ -537,12 +537,14 @@ Requires:       python%{python3_pkgversion}-cherrypy
...
 Requires:       python%{python3_pkgversion}-routes
 Requires:       python%{python3_pkgversion}-werkzeug
+Suggests:       python%{python3_pkgversion}-saml
 %endif
 %if 0%{?suse_version}
 Requires:       python%{python3_pkgversion}-CherryPy
 Requires:       python%{python3_pkgversion}-PyJWT
 Requires:       python%{python3_pkgversion}-Routes
 Requires:       python%{python3_pkgversion}-Werkzeug
+Suggests:       python%{python3_pkgversion}-python3-saml
```
Comment (tchaikov) [inline ceph.spec.in:547]: after reading the "Files" section in https://opensuse.pkgs.org/tumbleweed/opensuse-oss-aarch64/python2-python3-saml-1.9.0-1.1.noarch.rpm.html , i am convinced that this python2-python3-saml does not offer python3-saml.

File: ceph.spec.in [hunk 4]
```diff
@@ -537,12 +537,14 @@ Requires:       python%{python3_pkgversion}-cherrypy
 Requires:       python%{python3_pkgversion}-jwt
 Requires:       python%{python3_pkgversion}-routes
 Requires:       python%{python3_pkgversion}-werkzeug
+Suggests:       python%{python3_pkgversion}-saml
```
Comment (tchaikov) [inline ceph.spec.in:540]: could you file a bugzilla ticket asking for a epel8 branch for python3-saml ? since we have fc33 branch. see https://src.fedoraproject.org/rpms/python3-saml. i believe we can have epel8 with minimum efforts.

File: debian/control (hunk 1 above)
Comment (tchaikov) [reply debian/control:254]: i'd prefer drop this line. if we cannot offer this package.

Comment (tchaikov) [reply debian/control:254]: i am copying the discussion over IRC here

----

[20:38:23]  <epuertat__>	kefu, I prefer to 'track' that dependency in the package metadata. being a weak dependency it shouldn't block the installation if it's not found, and weak deps were created exactly for specifying optional components. IMHO it doesn't make sense to define it in the requirements.txt (where it's not optional) but not in the package spec
[20:39:29]  <kefu>	epuertat__ i think the reason we list a package in packaging is to help maintainer and enduser to consumer this software.
[20:41:24]  <kefu>	if we cannot offer them the package, and we don't plan to do so, why shall we list it in, for instance, d/control?
[20:43:11]  <kefu>	FWIW, podman is being ITP'ed.
[20:43:45]  <epuertat__>	kefu, I get your point, but if we follow that reasoning why we added python3-saml dependencies (xmlsec, ...) to the BuildRequires?
[20:44:16]  <kefu>	epuertat__ strictly speaking, they are not there.
[20:44:56]  <epuertat__>	kefu, by specifying the weak deps, we allow users to build those deps on their (or in the future if someone comes up with those)
[20:45:22]  <kefu>	because we use builddeps track build, make check, and runtime deps.
[20:46:36]  <kefu>	instead of scattering them around, i'd prefer put all of them in a single place. and guard them with either %if macro (in rpm spec) or comments in d/control.
[20:47:06]  <kefu>	i don't agree with "allow".
[20:47:33]  <epuertat__>	kefu, don't you mind if we follow this discussion in the PR, so we track any decision there?
[20:47:46]  <kefu>	if we want to help our user. a README or a doc is better.

Comment (epuertat) [reply debian/control:254]: Thanks. No problem, I will remove this line.

@ricardoasmarques , do you (SUSE) want this dependency also removed from ceph.spec?

File: ceph.spec.in (hunk 4 above)
Comment (epuertat) [reply ceph.spec.in:540]: @ktdreyer and @bigjust are working on [releasing some pkgs from Ken's Ceph copr to EPEL](https://trello.com/b/ZAMLb0pT/ceph-el8).

Comment (tchaikov) [reply ceph.spec.in:540]: those packages are mostly the dependencies of python3-cherrypy , and they do not contain python-saml. see https://copr.fedorainfracloud.org/coprs/ktdreyer/ceph-el8/packages/

Comment (epuertat) [reply ceph.spec.in:540]: Yup, I know. This whole thing comes from a chat Timothy A. Jeyasingh & I had yesterday with @ktdreyer at #ceph-dashboard Google Chat (that's why I prefer to keep discussions in shareable places :wink: ). Hope Ken doesn't mind I copy that here:

> In any regard, I think there may be no good reason why ceph.spec does not have Requires: python3-saml or Recommends: python3-saml
> It's quite likely an accident
> It is also important to verify that EPEL 8 has python3-saml in it
> Maybe that is why this was not done before
> I've been working with a newer member of our team (Justin Caratzas) to get all our dependencies in to EPEL 8
> I am thinking that python3-saml may not be on his list
> Justin is working from https://copr.fedorainfracloud.org/coprs/ktdreyer/ceph-el8/ , and tracking progress in https://trello.com/b/ZAMLb0pT/ceph-el8

File: ceph.spec.in (hunk 3 above)
Comment (epuertat) [reply ceph.spec.in]: This python3-saml package is really weird one, as it provides both python2- and python3-, but then the same author provides a separate python-saml that only works with Python 2.7... Nevertheless, I removed that weak dep and left only Fedora/CentOS'.

File: ceph.spec.in (hunk 4 above)
Comment (tchaikov) [reply ceph.spec.in:540]: i see. and i still suggest file a bz for creating the epel8 branch of python-saml. so it can be at least be tracked. and won't be forgotten over time.

File: ceph.spec.in [hunk 5]
```diff
@@ -537,6 +537,7 @@ Requires:       python%{python3_pkgversion}-cherrypy
...
```
Comment (tchaikov) [inline ceph.spec.in:540]: please use something like
```
%if 0%{?weak_deps}
Recommends:    python%{python3_pkgversion}-saml
%endif
```

we can drop the guards if we decide to drop the el7 support. but let's do this in a different PR if we really want to do this.

File: ceph.spec.in (hunk 4 above)
Comment (LenzGr) [reply ceph.spec.in:540]: Note that it was a deliberate decision to not add a dependency on python-saml to the spec file, as it's not required by default and the python code only requires the module when SSO is actively enabled.

Adding @ricardoasmarques to comment, as he's the original author of that feature.

Comment (LenzGr) [reply ceph.spec.in:540]: Having said that, I'm fine with adding it to the package dependencies in a "soft" form, e.g. as `Recommends`. Since the majority of users won't need this feature, I don't think we should complicate things unnecessarily.

File: ceph.spec.in [hunk 6]
```diff
@@ -537,6 +537,7 @@ Requires:       python%{python3_pkgversion}-cherrypy
...
 %endif
 %if 0%{?suse_version}
```
Comment (smithfarm) [inline ceph.spec.in:542]: If this is going to be added, can we do the same for SUSE, where the package name is `python3-python3-saml` (yes, confirmed, that is really the name)

i.e.

```
Recommends:    python%{python3_pkgversion}-python3-saml
```

here in this `suse_version` if block

Comment (tchaikov) [reply ceph.spec.in:542]: see also https://pkgs.org/download/python3-python3-saml

File: ceph.spec.in (hunk 4 above)
Comment (epuertat) [reply ceph.spec.in]: Thanks, @tchaikov & @LenzGr for your feedback!

@tchaikov there's already this [Bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1819302) (Ken and Timothy are working on that).

File: ceph.spec.in (hunk 5 above)
Comment (epuertat) [reply ceph.spec.in]: Done!

File: ceph.spec.in (hunk 6 above)
Comment (epuertat) [reply ceph.spec.in:544]: Thanks @smithfarm! Done.

**Comment by tchaikov** [comment]
https://shaman.ceph.com/repos/ceph/wip-kefu-testing-2020-04-08-1234/cf0ac110b4d3d57938d6df658c050472f6fc9ae9/

File: ceph.spec.in (hunk 4 above)
Comment (epuertat) [reply ceph.spec.in]: @tchaikov please let me know what you exactly mean, cause I don't follow. From my understanding, @ktdreyer & @bigjust were undertaking the addition of several Ceph deps. to EPEL8 (including this one and their deps) which will in turn involve the BZs to Fedora EPEL. What am I missing there? Thanks!

Comment (tchaikov) [reply ceph.spec.in]: discussed with @epuertat offline, my meant to suggest we need to track the efforts to package python3-saml in EPEL8, so our user can at least expect python3-saml when the package is found in the (weak) dependency list of ceph.

but since we are fully aware of this and we are going to package it for EPEL8, i am fine now =) 

thanks!

Comment (ktdreyer) [reply ceph.spec.in]: Who is packaging this for EPEL 8?

Comment (tchaikov) [reply ceph.spec.in]: i think Justin is on this. @epuertat, please correct me if i am wrong.

Comment (epuertat) [reply ceph.spec.in]: > Who is packaging this for EPEL 8?

@ktdreyer I initially understood @bigjust and you we porting stuff to EPEL8, but after our last discussion, Timothy agreed to take over on this.
